src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+
+        private async Task CloseAsyncCore(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
+        {
+            ThrowOnInvalidState(State,
+                WebSocketState.Open, WebSocketState.CloseReceived, WebSocketState.CloseSent);
+
+            WebSocketValidate.ValidateCloseStatus(closeStatus, statusDescription);
+
+            _tcsClose = new TaskCompletionSource<bool>();
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully connected or failed trying.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            // In the failure case we need to release the references and dispose of the objects.
+            using (cancellationToken.Register(() => _tcsClose.TrySetCanceled()))
+            {
+                _innerWebSocketCloseStatus = closeStatus;
+                _innerWebSocketCloseStatusDescription = statusDescription;
+                _innerWebSocket!.Invoke("close", (int)closeStatus, statusDescription);
+                await _tcsClose.Task.ConfigureAwait(continueOnCapturedContext: true);
+            }
+        }
+
+        public override Task CloseOutputAsync(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken) => throw new PlatformNotSupportedException();
+
+        private void ThrowIfNotConnected()
+        {
+            if (_state == _disposed)
+            {
+                throw new ObjectDisposedException(GetType().FullName);
+            }
+            else if (State != WebSocketState.Open)
+            {
+                throw new InvalidOperationException("WebSocket is not connected");
+            }
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (_state == _disposed)
+            {
+                throw new ObjectDisposedException(GetType().FullName);
+            }
+        }
+
+        private class ActionQueue<T>
+        {
+
+            private readonly SemaphoreSlim actionSem;
+            private readonly ConcurrentQueue<T> actionQueue;
+
+            public ActionQueue()
+            {
+                actionSem = new SemaphoreSlim(0);
+                actionQueue = new ConcurrentQueue<T>();
+            }
+
+            public void BufferPayload(T item)
+            {
+                actionQueue.Enqueue(item);
+                actionSem.Release();
+            }
+
+            public async Task<T> DequeuePayloadAsync(CancellationToken cancellationToken = default(CancellationToken))
+            {
+                while (true)
+                {
+                    await actionSem.WaitAsync(cancellationToken).ConfigureAwait(continueOnCapturedContext: true);
+
+                    T item;
+                    if (actionQueue.TryDequeue(out item))
+                    {
+                        return item;
+                    }
+                }
+            }
+        }
```

Comment:
Can you use a `Channel.CreateUnbuffered<T>()` instead of this `ActionQueue<T>`?  Then you'd just use its WriteAsync/TryWrite instead of BufferPayload and its ReadAsync instead of DequeuePayloadAsync.

Reply:
I was just going to suggest a channel 😄

Reply:
Yes, was going to look at all this Payload/ActionQueue implementation.  This was leftover from the mono implementation when all of this was implemented outside of the repo tree.  Now that it is incorporated and we can actually run tests against the implementation am quite sure this can all be refactored out.

Reply:
Addressed in recent commits